data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconBlockHandler.java

```diff
@@ -13,71 +13,116 @@
+import tech.pegasys.artemis.beaconrestapi.schema.BeaconBlockResponse;
 import tech.pegasys.artemis.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.artemis.provider.JsonProvider;
 import tech.pegasys.artemis.storage.ChainStorageClient;
 import tech.pegasys.artemis.storage.HistoricalChainData;
 
-public class BeaconBlockHandler implements BeaconRestApiHandler {
+public class BeaconBlockHandler implements Handler {
 
   private final ChainStorageClient client;
   private final HistoricalChainData historicalChainData;
+  public static final String ROUTE = "/beacon/block";
+  private final JsonProvider jsonProvider;
 
-  public BeaconBlockHandler(ChainStorageClient client, HistoricalChainData historicalChainData) {
+  public BeaconBlockHandler(
+      final ChainStorageClient client,
+      final HistoricalChainData historicalChainData,
+      final JsonProvider jsonProvider) {
     this.client = client;
     this.historicalChainData = historicalChainData;
+    this.jsonProvider = jsonProvider;
   }
 
+  @OpenApi(
+      path = ROUTE,
+      method = HttpMethod.GET,
+      summary = "Request that the node return the specified beacon chain block.",
+      tags = {"Beacon"},
+      queryParams = {
+        @OpenApiParam(name = "epoch", description = "Query by epoch number (uint64)"),
+        @OpenApiParam(name = "slot", description = "Query by slot number (uint64)"),
+        @OpenApiParam(name = "root", description = "Query by tree hash root (Bytes32)")
+      },
+      description =
+          "Request that the node return the beacon chain block that matches the provided criteria.",
+      responses = {
+        @OpenApiResponse(
+            status = "200",
+            content = @OpenApiContent(from = BeaconBlockResponse.class)),
+        @OpenApiResponse(status = "400", description = "Invalid parameter supplied"),
+        @OpenApiResponse(status = "404", description = "Specified block not found")
+      })
   @Override
-  public String getPath() {
-    return "/beacon/block";
-  }
+  public void handle(final Context ctx) throws Exception {
+    try {
+      if (ctx.queryParamMap().size() > 1) {
+        throw new IllegalArgumentException(
+            "Too many query parameters specified. Please supply only one.");
+      }
 
-  @Override
-  public Object handleRequest(RequestParams param) {
-    Map<String, List<String>> queryParamMap = param.getQueryParamMap();
-    if (queryParamMap.containsKey(ROOT)) {
-      Bytes32 root = Bytes32.fromHexString(param.getQueryParam(ROOT));
-      return client.getStore() != null ? client.getStore().getBlock(root) : null;
-    }
+      final Map<String, List<String>> queryParamMap = ctx.queryParamMap();
+      if (ctx.queryParamMap().containsKey(ROOT)) {
+        final Bytes32 root = Bytes32.fromHexString(validateQueryParameter(queryParamMap, ROOT));
+        if (client.getStore() != null) {
+          final SignedBeaconBlock block = client.getStore().getSignedBlock(root);
```

Comment:
so the store being null here is basically the SC_NO_CONTENT use-case that we have currently...

Reply:
Hrm i think this is probably a bigger problem... We need to probably use the combinedChainDataClient, because this client is only non finalized blocks which is something i found out doing beacon/state.
The upshot, is we can build most of the complexity of using chainDataClient v. historicalChainDataClient into the combinedChainDataClient, and then we're just responding to that.
The downside is then we really need to be respecting futures, because the storage is async.
The better thing might be to complete this task, then raise another to allow /beacon/block to retrieve finalized blocks...

Reply:
Agreed - best bet is to always use `CombinedChainDataClient` for REST APIs.  The `ChainStorageClient` only has non-finalised blocks, everything else is written to disk and dropped from memory so only available from `HistoricalChainDataClient`.  `CombinedChainDataClient` is designed to work out which one you need for you.

Separate PR is a sensible approach.   Javalin makes working with async data simple since you can pass a `Future` to `ctx.result`. `BeaconStateHandler` is an example of doing this.